content/en/docs/tasks/administer-cluster/kms-provider.md

```diff
@@ -90,57 +119,133 @@ You can develop a KMS plugin gRPC server using a stub file available for Go. For
+  results in a distinct response `(ciphertext, key_id, annotations)`.
+
+  If the plugin returns a non-empty `annotations` map, all map keys must be fully qualified domain names such as
+  `example.com`.
+
+  The API server does not perform the `EncryptRequest` procedure call at a high rate.  Plugin implementations should
+  still aim to keep each request's latency at under 100 milliseconds.
+
+* Decryption
+
+  The `DecryptRequest` procedure call provides the `(ciphertext, key_id, annotations)` from `EncryptRequest` and a UID
+  for logging purposes.  As expected, it is the inverse of the `EncryptRequest` call.  Plugins must verify that the
+  `key_id` is one that they understand - they must not attempt to decrypt data unless they are sure that it was
+  encrypted by them at an earlier time.
+
+  The API server may perform thousands of `DecryptRequest` procedure calls on startup to fill its watch cache.  Thus
+  plugin implementations must perform these calls as quickly as possible, and should aim to keep each request's latency
+  at under 10 milliseconds.
+
+* Understanding `key_id` and Key Rotation
+
+  The `key_id` is the public, non-secret name of the remote KMS KEK that is currently in use.  It may be logged
+  during regular operation of the API server, and thus must not contain any private data.  Plugin implementations
+  are encouraged to use a hash to avoid leaking any data.  The KMS v2 metrics take care to hash this value before
+  exposing it via the `/metrics` endpoint.
+
+  The API server considers the `key_id` returned from the `Status` procedure call to be authoritative.  Thus, a change
+  to this value signals to the API server that the remote KEK has changed, and data encrypted with the old KEK should
+  be marked stale when a no-op write is performed (as described below).  If an `EncryptRequest` procedure call returns a
+  `key_id` that is different from `Status`, the response is thrown away and the plugin is considered unhealthy.  Thus
+  implementations must guarantee that the `key_id` returned from `Status` will be the same as the one returned by
+  `EncryptRequest`.  Furthermore, plugins must ensure that the `key_id` is stable and does not flip-flop between values
+  (i.e. during a remote KEK rotation).
+
+  Plugins must not re-use `key_id`s, even in situations where a previously used remote KEK has been reinstated.  For
+  example, if a plugin was using `key_id=A`, switched to `key_id=B`, and then went back to `key_id=A` - instead of
+  reporting `key_id=A` the plugin should report some derivative value such as `key_id=A_001` or use a new value such
+  as `key_id=C`.
+
+  Since the API server polls `Status` about every minute, `key_id` rotation is not immediate.  Furthermore, the API
+  server will coast on the last valid state for about three minutes.  Thus if a user wants to take a passive approach
+  to storage migration (i.e. by waiting), they must schedule a migration to occur at `3 + N + M` minutes after the
```

Comment:
this question stems from my unfamiliarity of storage migration, and could be based on incomplete assumption. AIUI, the storage migration is based on the storageversion hash . Example below.
For storage migration for take effect, which object's storageVersionHash change would trigger the migration? Is `EncryptionConfiguration` GVK change trigger the migration. I understand details of the storage migration are beyond the scope of this documentation 
```
sh-4.2$ kubectl get --raw /api/v1 | jq '.resources[] | select(.name=="secrets")'
{
  "name": "secrets",
  "singularName": "",
  "namespaced": true,
  "kind": "Secret",
  "verbs": [
    "create",
    "delete",
    "deletecollection",
    "get",
    "list",
    "patch",
    "update",
    "watch"
  ],
  "storageVersionHash": "S6u1pOWzb84="
}
```

Reply:
Storage migration can be triggered either via the storage version migrator or a manual invocation of `kubectl get secrets --all-namespaces -o json | kubectl replace -f - ` to encrypt all existing Secrets with the new key or decrypt all if the `identity` provider is the first entry in the `EncryptionConfiguration` file.

Reply:
> AIUI, the storage migration is based on the storageversion hash

The `storageVersionHash` field is deprecated and will be removed in a future release.  It is also not aware of encryption at rest, and thus `storageVersionHash` has no meaning in relation to KMS v2.